jbi/retry.py

```diff
@@ -0,0 +1,61 @@
+import asyncio
+import logging
+from datetime import datetime, timedelta
+from os import getenv
+from time import sleep
+from typing import AsyncIterator
+
+import jbi.runner as runner
+from jbi.configuration import ACTIONS
+from jbi.queue import QueueItem, get_dl_queue
+
+CONSTANT_RETRY = getenv("CONSTANT_RETRY", "false") == "true"
+RETRY_TIMEOUT_DAYS = getenv("RETRY_TIMEOUT_DAYS", 7)
+
+
+async def retry_failed():
+    queue = get_dl_queue()
+    logger = logging.getLogger(__name__)
+    min_event_timestamp = datetime.now() - timedelta(days=int(RETRY_TIMEOUT_DAYS))
+
+    # load all bugs from DLQ
+    bugs: dict[int, AsyncIterator[QueueItem]] = await queue.retrieve()
+
+    for bugid, items in bugs.items():
+        prev_failed = False
+        async for item in items:
+            # skip if any previous retries for this bug have already failed
+            if prev_failed:
+                logger.info(
+                    "skipping event %s - previous events have failed for this bug",
+                    item.identifier,
+                )
+                continue
+
+            # skip and delete item if we have exceeded max_timeout
+            if item.timestamp < min_event_timestamp:
+                logger.warn("removing expired event %s", item.identifier)
+                await queue.done(item)
+                continue
+
+            try:
+                runner.execute_action(item.payload, ACTIONS)
+                await queue.done(item)
+            except Exception as ex:
```

Comment:
Should we split `runner.execute_action(item.payload, ACTIONS)` and `await queue.done(item)` into separate `try / except` blocks? There's a small but non-0 chance that we successfully execute an action, but then when we try to mark the item as done, something goes wrong.

We could even retry marking the item as done if something goes wrong.

Reply:
Mapping this out (this comment might get messy).

Current code workflow looks like this:
1. Process event successfully via `execute_action`
2. Fail to mark as done (can't delete file, likely IO error or maybe a k8s config issue)
3. Skip processing future events for this bug
4. On next retry, we reprocess the same event even though we processed it successfully the first time
5. Mark it as done successfully this time
6. Process other events in queue for this bug as normal

On the surface, that doesn't seem like a problem. We might process the same event twice (several hours apart), but ultimately we're updating the same data to the same values. What problems may occur here that I'm not thinking of?


If we alter the code to note that we failed marking the item as done it would look like:
1. Process event successfully via `execute_action`
2. Fail to mark as done
3. Update the existing file (rename or rewrite) to note that we failed to delete it. Or maybe write an additional file?
4. Continue processing other events for this bug since we didn't actually fail to update
5. On next retry, check for this note so we know to skip `execute_action` and just try to mark it as done
6. Continue as normal

I think this adds more complexity (more branching, more test cases) and is likely to run into the same IO issue that prevented us from deleting the file. I don't think we gain anything, unless we might stop ourselves from overwriting something in Jira that a user would have changed manually.

Does that stream of thoughts seem right? Am I missing anything?

Reply:
> What problems may occur here that I'm not thinking of?

Some operations are creation, like comments for example. 
We may end up with a comment being posted twice. I don't think it's critical and I value simple code over this (especially because once this is setup, failing to write on a disk should not happen easily)

Reply:
That's a good thing to note. If that becomes an issue, I think the easiest solution is to do upserts for all create/update commands. That might be a little cumbersome through the Jira API but hopefully we can create a simple wrapper for ourselves if it is.

This also fixes the issue where something was created, but didn't make it into Jira (for whatever reason). And then we receive an update for it later.